lib/viewmodels/upvote_view_model.dart

```diff
@@ -0,0 +1,97 @@
+/// user triggers an upvote or downvote action.
+///
+/// This view model must be AutoDisposed because it is created on the fly
+/// for each post that is displayed on the screen.
+/// Thus, when the post is removed from the screen, the view model is disposed.
+class UpVoteViewModel extends AutoDisposeFamilyAsyncNotifier<PostVote,
+    ({PostIdFirestore postId})> {
+  @override
+  FutureOr<PostVote> build(({PostIdFirestore postId}) arg) async {
+    final postId = arg.postId;
+    final uid = ref.watch(uidProvider);
+    if (uid == null) {
+      throw Exception("User is not logged in");
+    }
+
+    final voteRepository = ref.watch(postUpvoteRepositoryProvider);
+    final postRepository = ref.watch(postRepositoryProvider);
+
+    final postFuture = postRepository.getPost(postId);
+    final upvoteStateFuture = voteRepository.getUpvoteState(uid, postId);
+
+    // Await the futures in parallel for better responsiveness
+    final results = await Future.wait([postFuture, upvoteStateFuture]);
+    final post = results[0] as PostFirestore;
+    final upvoteState = results[1] as UpvoteState;
+
+    return PostVote(
+      upvoteState: upvoteState,
+      votes: post.data.voteScore,
+    );
+  }
+
+  /// This method is used to perform the upvote action of a post.
+  /// If the post is already upvoted, it will remove the upvote.
+  Future<void> triggerUpVote() => _triggerVote(UpvoteState.upvoted);
+
+  /// This method is used to perform the downvote action of a post.
+  /// If the post is already downvoted, it will remove the downvote.
+  Future<void> triggerDownVote() => _triggerVote(UpvoteState.downvoted);
+
+  Future<void> _triggerVote(UpvoteState selectedUpVoteState) async {
+    final postId = arg.postId;
+    final uid = ref.watch(uidProvider);
+    if (uid == null) {
+      throw Exception("User is not logged in");
+    }
+
+    final voteRepository = ref.read(postUpvoteRepositoryProvider);
+
+    final currPostVote = await future;
+    final currUpVoteState = currPostVote.upvoteState;
+
+    // If the user trigger the same vote state, we remove the vote.
+    // (For instance, double clicking the upvote button should add then remove the upvote)
+    final newVoteState = currUpVoteState == selectedUpVoteState
+        ? UpvoteState.none
+        : selectedUpVoteState;
+
+    final newVotes =
+        currPostVote.votes + newVoteState.increment - currUpVoteState.increment;
+
+    // Here, we update the state manually because we only want to update the state
+    // locally without fetching the data from the database again.
+    state = AsyncData(
+      PostVote(
+        upvoteState: newVoteState,
+        votes: newVotes,
+      ),
+    );
+
+    await voteRepository.setUpvoteState(uid, postId, newVoteState);
+  }
+}
+
+final postVoteProvider = AsyncNotifierProvider.autoDispose
+    .family<UpVoteViewModel, PostVote, ({PostIdFirestore postId})>(
+  UpVoteViewModel.new,
```

Comment:
Here the code should not use this `.new` syntax as it was an old Dart way of creating objects that is deprecated today.

Reply:
Oh, I didn't know this was a deprecated feature.
Thanks for pointing it out !
Fixed in 5b00f1abc10da9c397cb53ac476a1681227abf8a.